### PR TITLE
[micromega] native support for boolean operators

### PR DESCRIPTION
**Pull request by fajb**
* `lia` supports boolean comparisons and operators
    + Z.{eqb, leb, ltb, geb, gtb}
    + andb, orb, implb, eqb, negb, ||
* `zify` has been modified to work in harmony
    + `ZifyBool.v` injects comparison over `nat` to comparisons over `Z`  

**Kind:** feature 
(this is a clean revival of #11047 )

- [x] Added / updated test-suite
- [x] Corresponding documentation was added /
- [x] Entry added in the changelog

### CODE REVIEW

**Comment by fajb** [comment]
CI passed. The windows failures seem unrelated.

**Review by maximedenes — CHANGES_REQUESTED** [review]
IIUC, this PR is introducing more hardcoded paths to constants in micromega. I'd prefer that we first switch to the `Coqlib.lib_ref` API, or at least that we use it for the newly needed constants.

**Comment by fajb** [comment]
@maximedenes fair enough. I'll make the modifications as soon as PR #11990 is merged.

**Comment by Zimmi48** [comment]
@fajb Do you think you'll finish this in time for the 8.12 branching (on Friday). Maybe @coq/micromega-maintainers can still review this now even before the overlays are ready to speed the process.

**Comment by fajb** [comment]
@Zimmi48, I hope it can be ready.

All failures should be solved by PR that were very quickly merged yesterday.
There is still a single overlay for cross_crypto.
PS: I'll restart the CI in a few hours...

**Comment by fajb** [comment]
Current status:
- cross_crypto needs PR https://github.com/adampetcher/fcf/pull/33 
- bedrocks  needs to upgrade kami
- vst is spurious
- fiat_crypto has timed out (but succeeded earlier...)

I'll restart the CI when possible.

**Comment by fajb** [comment]
@denes , the PR is now using `Coqlib.lib_ref`.
@vbgl or maybe @thery . I would be glad if you can spare some time to review. Thanks.

**Review by vbgl — COMMENTED** [review]
It might be worth mentioning in the changelog that since micromega tactics are more powerful, proof scripts that expect them to fail may break.

File: theories/micromega/Tauto.v
Comment (fajb) [reply theories/micromega/Tauto.v]: Done. That was more intrusive than expected :(

**Comment by fajb** [comment]
@vbgl I also updated the changelog.

**Comment by vbgl** [comment]
Maxime, do you have a better opinion of this PR now?

Frédéric, are there open PRs corresponding to the overlays?

**Comment by fajb** [comment]
@vbgl , no PR for the  overlays. They simply  update the versions of kami for bedrock2 and fcf for cross_crypto. The modifications of kami and fcf are already merged.
If that's the way to go, I can open a PR...

**Comment by vbgl** [comment]
Yes, I’ve done this in the past. See e.g., https://github.com/mit-plv/bedrock2/pull/104

**Comment by fajb** [comment]
Thanks @vbgl.  I have opened PR https://github.com/mit-plv/cross-crypto/pull/17 and PR https://github.com/mit-plv/bedrock2/pull/158

**Comment by vbgl** [comment]
Thanks. Do you intend to add something to the test-suite?

**Comment by fajb** [comment]
No. But, I have a few micro tests that I did not commit.

**Comment by ejgallego** [comment]
@coq/micromega-maintainers would you like this to be shipped in 8.12 ? We are going to branch very soon, but we can accommodate a couple of days if you are positive you want this in.

**Comment by fajb** [comment]
@ejgallego, from my side,  I would be glad if it would ship...

**Comment by ejgallego** [comment]
> I would be glad if it would ship

Looks reasonable to me; I'm doing a re-ping so @maximedenes can update his review as he may not have seen @vbgl message.

**Review by vbgl — APPROVED** [review]

File: theories/micromega/ZifyBool.v
Comment (pi8027) [inline theories/micromega/ZifyBool.v:67]:
```suggestion
  { TBOp := Z.eqb ; TBOpInj := ltac:(reflexivity)}.
Add BinOp Op_Zeqb.
```

File: theories/micromega/ZifyBool.v
Comment (pi8027) [inline theories/micromega/ZifyBool.v:84]:
```suggestion
  { TBOp := Z.ltb ;
    TBOpInj := ltac:(reflexivity)
  }.
Add BinOp Op_Zltb.
```

**Review by pi8027 — CHANGES_REQUESTED** [review]

**Comment by pi8027** [comment]
Hello. I discovered the following issues in this PR.

1. This change makes the following script does not work. I guess that `lia` should have an extra constraint `b = true \/ b = false` here.
   ```coq
   Require Import Lia ZifyBool.
   Goal forall b : bool, negb (negb b) = b. Proof. lia. Qed.
   ```
2. `UnOpSpec` and `BinOpSpec` do not support functions that have different input and output types, such as `Z.b2z` and `Nat.b2n`.

**Comment by fajb** [comment]
Hi @pi8027 , thanks for the feedback.
For 1., there is easy to fix but I fear a case-split blow-up -- I'll add the constraint anyway.
For 2., that requires a little work. I was lazy and did not think about such operators.

**Comment by fajb** [comment]
@pi8027 thinking twice, the constraint is not enough.
Currently, lia does not deal with boolean variables. The goal is out-of-scope.
This is a regression: the previous encoding made it possible as bool was injected into Z

**Comment by ejgallego** [comment]
> This is a regression: the previous encoding made it possible as bool was injected into Z

Unfortunately we have to branch now :( Two change requests seem a bit too much given that beta is expected in one week sorry. Hopefully we can get this in very soon in 8.13

**Comment by fajb** [comment]
@pi8027 I have updated ZiyBool. I believe the mapping bool -> bool is the right way forward. Do deal with boolean variables, I propose to eliminate them using an explicit Ltac case-split.
The last commit should solve both your issues. Tell me if/when you uncover other issues.

**Comment by pi8027** [comment]
I'm sorry for blocking this PR. Now it almost works as I expected! But also another issue puzzles me in [mczify](https://github.com/pi8027/mczify).

The problem is that replacing `{| ... |}` (the constructor syntax for records) with `{ ... }` (that for class instances) in [1] makes a `Qed` command fails with the following message in [2].
```coq
Error: Illegal application: 
The term "mkapp2" of type
 "forall (S1 S2 S3 T1 T2 T3 : Type) (Op : S1 -> S2 -> S3) (I1 : S1 -> T1) (I2 : S2 -> T2)
    (I3 : S3 -> T3) (TBOP : T1 -> T2 -> T3),
  (forall (n : S1) (m : S2), I3 (Op n m) = TBOP (I1 n) (I2 m)) ->
  forall (s1 : S1) (t1 : T1),
  I1 s1 = t1 -> forall (s2 : S2) (t2 : T2), I2 s2 = t2 -> I3 (Op s1 s2) = TBOP t1 t2"
cannot be applied to the terms
 "int" : "Set"
 "int" : "Set"
 "GRing.Ring.sort int_Ring" : "Type"
 "Z" : "Set"
 "Z" : "Set"
 "Z" : "Set"
 "divz" : "int -> int -> int_Ring"
 "Z_of_int" : "int -> Z"
 "Z_of_int" : "int -> Z"
 "Z_of_int" : "int -> Z"
 "divZ" : "Z -> Z -> Z"
 "Op_divz_subproof" : "forall n m : int, Z_of_int (n %/ m)%Z = divZ (Z_of_int n) (Z_of_int m)"
 "int_of_Z n" : "int"
 "(id : Z -> Z) n" : "Z"
 "mkapp Z int Z Z int_of_Z id Z_of_int (id : Z -> Z) int_of_ZK n n (erefl n)"
   : "Z_of_int (int_of_Z n) = n"
 "int_of_Z m" : "int"
 "(id : Z -> Z) m" : "Z"
 "mkapp Z int Z Z int_of_Z id Z_of_int (id : Z -> Z) int_of_ZK m m (erefl m)"
   : "Z_of_int (int_of_Z m) = m"
The 3rd term has type "Type@{GRing.Ring.type.u0}" which should be coercible to 
"Type@{mkapp2.u2}".
```
[1] https://github.com/pi8027/mczify/blob/a3bf8b502e65db7748c25cf4540f80992b1d5af1/theories/zify.v#L347-L348
[2] https://github.com/pi8027/mczify/blob/a3bf8b502e65db7748c25cf4540f80992b1d5af1/theories/zify.v#L384

Since the `{ ... }` syntax seems to enable an elaboration mechanism specialized for type classes, I sometimes prefer to use it rather than `{| ... |}`, but it also seems to lead us to universe inconsistencies. Does the above failure also mean that the `lia` tactic does not perform type checking sufficiently?

**Comment by fajb** [comment]
@pi8027 , blocking now means less patching later. No worry. This is just too bad, it does not make it into the next release.
I have to admit, `zify` ignores universes. Naively, as `zify` is about some fragment of first-order logic, I never thought this would matter. 
I don't understand precisely what's going on...  and why { } vs {|  |} makes the difference.

**Comment by pi8027** [comment]
@fajb I see that using `{ ... }` makes an implicit argument `GRing.Ring.sort int_Ring` rather than `int`. So I think I should use `{| ... |}` here anyway, but it does not explain the above error.

**Comment by pi8027** [comment]
Also, the master branch has the same issue. Since I found a workaround, I'm fine to merge this PR without fixing it and to investigate/fix it later.

**Comment by fajb** [comment]
@pi8027  I hope that only people adding declarations can be bitten. 
What I find a bit troubling is that the type of `divz` has type `int -> int -> GRing.Ring.sort int_Ring`.
Forcing the type (int -> int -> int) solves the issue.

File: plugins/micromega/g_zify.mlg
Comment (pi8027) [inline plugins/micromega/g_zify.mlg:31]: Shouldn't this deprecation of `Add Spec` be documented in CHANGELOG?

File: theories/micromega/QMicromega.v
Comment (pi8027) [inline theories/micromega/QMicromega.v:4]: I think this change should be reverted.

File: theories/micromega/RMicromega.v
Comment (pi8027) [inline theories/micromega/RMicromega.v:4]: Same as above.

File: theories/micromega/ZifyBool.v
Comment (pi8027) [inline theories/micromega/ZifyBool.v:25]: Does `TBOpInj := fun _ _ => eq_refl` make this more robust?

**Review by pi8027 — COMMENTED** [review]
I think I'm not the right person to review the plugin part of this PR, but now it works and it is fine to me as a user, except a few minor issues and remarks.

File: plugins/micromega/g_zify.mlg
Comment (pi8027) [reply plugins/micromega/g_zify.mlg:31]: Actually `micromega.rst` has not been updated accordingly.

**Comment by pi8027** [comment]
The current way to deal with boolean variables is still incomplete:
```coq
From Coq Require Import Lia ZifyBool.

Goal forall (f : nat -> bool), negb (negb (f 0)) = f 0.
Proof. lia. Qed.
```

File: plugins/micromega/g_zify.mlg
Comment (fajb) [reply plugins/micromega/g_zify.mlg:31]: The `Add xxx` are currently undocumented...

File: theories/micromega/QMicromega.v
Comment (fajb) [reply theories/micromega/QMicromega.v]: done.

File: theories/micromega/RMicromega.v
Comment (fajb) [reply theories/micromega/RMicromega.v]: done.

File: theories/micromega/ZifyBool.v
Comment (fajb) [reply theories/micromega/ZifyBool.v]: done.

**Comment by fajb** [comment]
This is fixed by updating the `zify_post_hook`.

> ```coq
> Require Import Lia ZifyBool.
> Goal forall (f : nat -> bool), negb (negb (f 0)) = f 0. Proof. lia. Qed.
> ```

File: plugins/micromega/g_zify.mlg
Comment (pi8027) [reply plugins/micromega/g_zify.mlg:31]: Sure. Actually what I saw in the documentation was `Show Zify ...` commands.

**Review by pi8027 — APPROVED** [review]
I'm quite happy with these changes. Thanks @fajb!

**Comment by fajb** [comment]
@pi8027 FYI, I have deprecated the Add <X> command to be replaced by Add Zify <X>. 
(This is documented and avoids a clash with the existing Add command).

**Comment by fajb** [comment]
@maximedenes is-there still something blocking this PR?

**Comment by pi8027** [comment]
Ping @vbgl and @maximedenes.

@fajb Also, it may be worth rebasing this branch to refresh CI.

**Comment by maximedenes** [comment]
Sorry, the last messages went below my radar. I'll rebase so that we get a new CI run.

**Review by maximedenes — APPROVED** [review]
LGTM, sorry for the delay!

**Comment by maximedenes** [comment]
I believe the metacoq failure is known and unrelated.

**Pull request merged**
